File: sdks/python/reshadx/resources/risk.py
"""ReshADX Python SDK - Risk Resource"""

from typing import Any, Dict, List, Optional

from ..models.risk import DeviceFingerprint, RiskAssessment, SimSwapCheck
from ..utils.http import HttpClient


class Risk:
    """Risk resource"""

    def __init__(self, http: HttpClient):
        self.http = http

    def assess(
        self,
        amount: int,
        account_id: str,
        device_fingerprint: Dict[str, Any],
    ) -> Dict[str, RiskAssessment]:
        """Assess risk for a transaction"""
        data = {
            "amount": amount,
            "accountId": account_id,
            "deviceFingerprint": device_fingerprint,
        }

        response = self.http.post("/risk/assess", json_data=data)
        return {"assessment": RiskAssessment(**response["assessment"])}

    def check_sim_swap(self, device_fingerprint: Dict[str, Any]) -> SimSwapCheck:
        """Check for SIM swap fraud"""
        data = {"deviceFingerprint": device_fingerprint}

        response = self.http.post("/risk/sim-swap/check", json_data=data)
        return SimSwapCheck(**response)

    def get_alerts(
        self,
        status: Optional[str] = None,
        risk_level: Optional[str] = None,
        page: int = 1,
        limit: int = 50,
    ) -> Dict[str, Any]:
        """Get fraud alerts for current user"""
        params: Dict[str, Any] = {"page": page, "limit": limit}

        if status:
            params["status"] = status
        if risk_level:
            params["riskLevel"] = risk_level

        return self.http.get("/risk/alerts", params=params)

    def report_fraud(
        self,
        transaction_id: str,
        reason: str,
        details: Optional[str] = None,
    ) -> Dict[str, str]:
        """Report a transaction as fraud"""
        data = {
            "transactionId": transaction_id,
            "reason": reason,
        }

        if details:
            data["details"] = details

        return self.http.post("/risk/fraud/report", json_data=data)

    def get_device_trust_score(self, device_fingerprint: Dict[str, Any]) -> Dict[str, Any]:
        """Get device trust score"""
        data = {"deviceFingerprint": device_fingerprint}
        return self.http.post("/risk/device/trust-score", json_data=data)

    def get_velocity_checks(self) -> Dict[str, Any]:
        """Get velocity checks (transaction frequency analysis)"""
        return self.http.get("/risk/velocity-checks")

    def whitelist_device(
        self,
        device_id: str,
        device_name: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Whitelist a device"""
        data = {"deviceId": device_id}

        if device_name:
            data["deviceName"] = device_name

        return self.http.post("/risk/device/whitelist", json_data=data)

    def remove_device_whitelist(self, device_id: str) -> Dict[str, bool]:
        """Remove device from whitelist"""
        return self.http.delete(f"/risk/device/whitelist/{device_id}")

File: sdks/typescript/src/resources/risk.ts
/**
 * ReshADX TypeScript SDK - Risk Resource
 */

import { HttpClient } from '../utils/http';
import {
  AssessRiskRequest,
  RiskAssessment,
  SimSwapCheckRequest,
  SimSwapCheckResponse,
  DeviceFingerprint,
} from '../types';

export class Risk {
  constructor(private http: HttpClient) {}

  /**
   * Assess risk for a transaction
   */
  async assess(request: AssessRiskRequest): Promise<{ assessment: RiskAssessment }> {
    return this.http.post('/risk/assess', request);
  }

  /**
   * Check for SIM swap fraud
   */
  async checkSimSwap(request: SimSwapCheckRequest): Promise<SimSwapCheckResponse> {
    return this.http.post<SimSwapCheckResponse>('/risk/sim-swap/check', request);
  }

  /**
   * Get fraud alerts for current user
   */
  async getAlerts(params: {
    status?: 'PENDING' | 'REVIEWED' | 'RESOLVED';
    riskLevel?: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
    page?: number;
    limit?: number;
  } = {}): Promise<{
    alerts: Array<{
      alertId: string;
      type: string;
      riskLevel: string;
      description: string;
      status: string;
      createdAt: string;
    }>;
    pagination: {
      total: number;
      page: number;
      limit: number;
      hasMore: boolean;
    };
  }> {
    return this.http.get('/risk/alerts', { params });
  }

  /**
   * Report a transaction as fraud
   */
  async reportFraud(transactionId: string, reason: string, details?: string): Promise<{
    reportId: string;
    status: string;
  }> {
    return this.http.post('/risk/fraud/report', {
      transactionId,
      reason,
      details,
    });
  }

  /**
   * Get device trust score
   */
  async getDeviceTrustScore(deviceFingerprint: DeviceFingerprint): Promise<{
    trustScore: number;
    riskLevel: 'LOW' | 'MEDIUM' | 'HIGH';
    factors: string[];
  }> {
    return this.http.post('/risk/device/trust-score', { deviceFingerprint });
  }

  /**
   * Get velocity checks (transaction frequency analysis)
   */
  async getVelocityChecks(): Promise<{
    dailyTransactionCount: number;
    dailyTransactionAmount: number;
    weeklyTransactionCount: number;
    weeklyTransactionAmount: number;
    unusualActivity: boolean;
    flags: string[];
  }> {
    return this.http.get('/risk/velocity-checks');
  }

  /**
   * Whitelist a device
   */
  async whitelistDevice(deviceId: string, deviceName?: string): Promise<{
    success: boolean;
    deviceId: string;
  }> {
    return this.http.post('/risk/device/whitelist', { deviceId, deviceName });
  }

  /**
   * Remove device from whitelist
   */
  async removeDeviceWhitelist(deviceId: string): Promise<{ success: boolean }> {
    return this.http.delete(`/risk/device/whitelist/${deviceId}`);
  }
}
